src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1079,30 +1080,49 @@ MonodroidRuntime::monodroid_dlopen_log_and_return (void *handle, char **err, con
 		delete[] full_name;
 	}
 
+	if (!need_api_init)
+		return handle;
+
+	auto api = new MonodroidExternalAPI_Impl ();
```

Comment:
I believe this is a possible memory leak: there's a `new` here, but no corresponding `delete external_api` anywhere in this PR.

If I understand the scenario correctly, *every time* we `dlopen("libxa-mono-api.so")`, Android will *clear* the memory used for `static`/global variables of that native lib (the BSS section), and without some form of `_fini()` in `libxa-mono-api.so` -- assuming Bionic even supports `_init`/`_fini`/`__attribute((destructor))`/etc.  (See also the [**dlopen**(3) man page](https://linux.die.net/man/3/dlopen)?).

Without a way to free the instance that `libxa-mono-api.so!external_api` refers to, every *subsequent* `dlopen("libxa-mono-api.so")` will thus leak the `MonodroidExternalAPI_Impl` allocated from the *previous* `dlopenI("libxa-mono-api.so")` invocation.

A possible fix is to *avoid* `new`/`delete`, and instead use a global:

```cc
static MonodroidExternalAPI_Impl _monodroid_ext_api;

void*
MonodroidRuntime::monodroid_dlopen_log_and_return (…)
{
    …
    if (!api_init (&_monodroid_ext_api)) {
        …
    }
    …
}
```

This way we only ever create *one* instance of `MonodroidExternalAPI_Impl`, and it's never leaked.

Reply:
This isn't a safe way. Yes, the bug is that the memory is cleared on this particular Android emulator image (and the scenario was that an `armeabi` app was running on an `x86` emulator). But there's no guarantee that this will always be the case (in fact, the app I tested before our `HelloWorld` sample, did reload the DSO, just rechecked it, and return a different value of its handle) and, even though we don't have any state in the API class **now**, I'd rather leak a tiny amount of memory than risk sharing state with an unknown number of the `DSO` instances.

Reply:
I will, however, investigate what bionic supports wrt destructors

Reply:
How often is this codepath hit?  How much memory could we possibly leak?

Are we hitting this codepath *every time* we *invoke* a P/Invoke to a `__Internal` method, in which case we could be hitting this code path *thousands*-*millions* of times, which could be a significant memory leak?

Or do we hit this code path *every time* we *resolve* a P/Invoke to a `__Internal` method, in which case we could *at most* leak dozens of instances, *maybe*?  (Though each instance is probably on a separate code page, so that could still add up…)

What's our worst-case scenario here?

Reply:
> But there's no guarantee that this will always be the case…

I don't understand this response.

In the solution of this PR, we have two libraries in the `.apk` of interest:

  * `libmonodroid.so`
  * `libxa-mono-api.so`

We're setting things up such that we *only ever* load `libmonodroid.so` *once*, by using `System.loadLibrary("monodroid")` in the *Java* bootstrap code.  All C/C++ `dlopen()` invocations, meanwhile, *never* load `libmonodroid.so`, but instead *always* load `libxa-mono-api.so`.

In this setup, the suggested `_monodroid_ext_api` would be loaded in `libmonodroid.so`.  Thus, the object would only ever be created once, and would never move around in memory, and would never have it's BSS cleared.

The loading of `libxa-mono-api.so`, meanwhile, would *itself* contain only *one* pointer, which would be "reset" whenever we `dlopen("libxa-mono-api.so")` via our invocation of `libxa-mono-api.so!_monodroid_init_external_api`.

> in fact, the app I tested before our HelloWorld sample, did reload the DSO, just rechecked it, and return a different value of its handle

*Which* DSO was reloaded?  I'm *assuming* you mean `libxa-mono-api.so`, right?  So what would it matter if the address returned by `dlopen("libxa-mono-api.so")` differs?  We don't *need* to allocate a different `MonodroidExternalAPI_Impl` instance *per `libxa-mono-api.so` loaded*; all loaded `libxa-mono-api.so` libraries can all use the *same* instance, no?

> I'd rather leak a tiny amount of memory than risk sharing state with an unknown number of the DSO instances.

Given the nature of things, I'm not sure that there's much risk to this sharing of state.

Reply:
> > But there's no guarantee that this will always be the case…
> 
> I don't understand this response.
> 
> In the solution of this PR, we have two libraries in the `.apk` of interest:
> 
>     * `libmonodroid.so`
> 
>     * `libxa-mono-api.so`
> 
> 
> We're setting things up such that we _only ever_ load `libmonodroid.so` _once_, by using `System.loadLibrary("monodroid")` in the _Java_ bootstrap code. All C/C++ `dlopen()` invocations, meanwhile, _never_ load `libmonodroid.so`, but instead _always_ load `libxa-mono-api.so`.
> In this setup, the suggested `_monodroid_ext_api` would be loaded in `libmonodroid.so`. Thus, the object would only ever be created once, and would never move around in memory, and would never have it's BSS cleared.
> The loading of `libxa-mono-api.so`, meanwhile, would _itself_ contain only _one_ pointer, which would be "reset" whenever we `dlopen("libxa-mono-api.so")` via our invocation of `libxa-mono-api.so!_monodroid_init_external_api`.

1. Mono asks us to resolve `__Internal`, we dlopen the DSO, initialize it with the API, return the handle
2. The DSO somehow affects "state" of the API (e.g. one of the methods there needs to store some state and it's modified as a result of some call)
3. Mono asks us to resolve `__Internal` again (for instance for a different code path which hasn't been hit before), so we repeat 1. and 2.
4. *Both* instances created above are used by Mono at different times, stepping on each other's toes by modifying the state of the shared API class

> 
> > in fact, the app I tested before our HelloWorld sample, did reload the DSO, just rechecked it, and return a different value of its handle
> 
> _Which_ DSO was reloaded? I'm _assuming_ you mean `libxa-mono-api.so`, right? So what would it matter if the address returned by `dlopen("libxa-mono-api.so")` differs? We don't _need_ to allocate a different `MonodroidExternalAPI_Impl` instance _per `libxa-mono-api.so` loaded_; all loaded `libxa-mono-api.so` libraries can all use the _same_ instance, no?

Yes, `libxa-mono-api.so`, since this is the "DSO" featured by this PR. If the address differs, then we have TWO or more copies of the assembly, all using a SINGLE instance of the API which *may* have some state. the fact it doesn't have anything now doesn't matter - in the future we may introduce some state and get bitten by the shared nature of the class  if there are multiple instances of DSOs using it.

> 
> > I'd rather leak a tiny amount of memory than risk sharing state with an unknown number of the DSO instances.
> 
> Given the nature of things, I'm not sure that there's much risk to this sharing of state.

I disagree, as replied above.

Reply:
> How often is this codepath hit? How much memory could we possibly leak?
> 
> Are we hitting this codepath _every time_ we _invoke_ a P/Invoke to a `__Internal` method, in which case we could be hitting this code path _thousands_-_millions_ of times, which could be a significant memory leak?
> 
> Or do we hit this code path _every time_ we _resolve_ a P/Invoke to a `__Internal` method, in which case we could _at most_ leak dozens of instances, _maybe_? (Though each instance is probably on a separate code page, so that could still add up…)
> 
> What's our worst-case scenario here?

From what I see, the `__Internal` DSO lookup is done once per assembly, so we'd leak N times where N equals the number of assemblies.

Reply:
Next question is *how much is leaked*?  Are we leaking a byte?  A *page*?  Multiple pages?  Unfortunately, this is likely based on **malloc**(3) behavior, and I don't know offhand what it's likely to do. :-/

We "only" likely to have app in the low-hundreds of assemblies (scary, but pull in Forms & AndroidX & NuGet and…), so if it's only once per assembly, this could either be an ignorable amount (1 byte per allocation) or a shockingly large amount (1 or more pages per allocation).

Maybe a custom allocator would be appropriate?

Reply:
I don't think it's once per *every* assembly - it's more likely per assembly which has any `__Internal` p/invokes which right now is two assemblies (?) in our collection. Size of the class is 8 bytes. Bionic uses `jemalloc` which will round up the allocation size to at least `sizeof(double)` and the allocated objects are packed closely together avoid wasting space (http://jemalloc.net/jemalloc.3.html - see the `IMPLEMENTATION NOTES` section) so I really think we shouldn't go overboard with custom allocator and that we don't have too much to worry about in this regard